File: Sources/macOS/Sources/OnScreen/Features/LeavesTracesWhileWalking.swift
import Foundation
import Yage

class LeavesTracesWhileWalking: Capability {
    var speciesForTraceEntities: Species!
    var timeBetweenSpawns: TimeInterval = 5
    var traceExpirationTime: TimeInterval = 1
    
    private var traces: [Entity] = []
    private var lastSpawnDate: Date {
        traces.last?.creationDate ?? Date(timeIntervalSince1970: 0)
    }
    
    override func install(on subject: Entity) {
        isEnabled = false
        super.install(on: subject)
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) { [weak self] in
            self?.isEnabled = true
        }
    }
    
    override func doUpdate(with collisions: Collisions, after time: TimeInterval) {
        if shouldSpawnTrace() {
            spawnTrace()
        }
        removeExpiredTraces()
    }
    
    override func kill(autoremove: Bool = true) {
        super.kill(autoremove: autoremove)
        traces.forEach { $0.kill() }
        traces.removeAll()
    }
    
    func shouldSpawnTrace() -> Bool {
        guard subject?.state == .move else { return false }
        guard abs(lastSpawnDate.timeIntervalSinceNow) > timeBetweenSpawns else { return false }
        return true
    }
    
    private func spawnTrace() {
        guard let species = speciesForTraceEntities else { return }
        guard let subject, let world = subject.world else { return }
        let entity = Entity(
            species: species,
            id: LeavesTracesWhileWalking.nextTraceId(),
            frame: subject.frame.offset(by: traceOffset()),
            in: world
        )
        entity.isEphemeral = true
        world.children.append(entity)
        traces.append(entity)
    }
    
    func traceOffset() -> CGPoint {
        .zero
    }
    
    private func shouldRemoveTrace() -> Bool {
        lastSpawnDate.timeIntervalSinceNow > traceExpirationTime
    }
    
    private func removeExpiredTraces() {
        traces
            .filter { abs($0.creationDate.timeIntervalSinceNow) > traceExpirationTime }
            .forEach { remove(trace: $0) }
    }
    
    private func remove(trace: Entity) {
        trace.kill()
        traces.remove(trace)
    }
    
    private static func nextTraceId() -> String {
        incrementalId += 1
        return "Trace-\(incrementalId)"
    }
    
    private static var incrementalId = 0
}

File: Sources/windows/onscreen/features/leaves_traces_while_walking.py
from datetime import datetime
from yage.models.capability import Capability
from yage.models.entity import Entity
from yage.models.entity_state import EntityState
from yage.utils.geometry import Point


class LeavesTracesWhileWalking(Capability):
    def __init__(self, subject):
        super().__init__(subject)
        self.species_for_trace_entities = None
        self.time_between_spawns = 5
        self.trace_expiration_time = 1
        self.traces = []
        self.is_enabled = True

    def do_update(self, collisions, time):
        if self.should_spawn_trace():
            self.spawn_trace()
        self.remove_expired_traces()

    def kill(self, autoremove=True):
        super().kill(autoremove)
        for trace in self.traces:
            trace.kill(autoremove)
        self.traces = []

    def should_spawn_trace(self):
        if self.subject.state != EntityState.MOVE:
            return False
        if self._seconds_since(self.last_spawn_date()) <= self.time_between_spawns:
            return False
        return True

    def last_spawn_date(self):
        if not self.traces:
            return datetime.fromtimestamp(0)
        return self.traces[-1].creation_date

    def spawn_trace(self):
        if not self.species_for_trace_entities:
            return
        if not self.subject:
            return
        if not self.subject.world:
            return

        entity = Entity(
            species=self.species_for_trace_entities,
            id=LeavesTracesWhileWalking.next_trace_id(),
            frame=self.subject.frame.offset(point=self.trace_offset()),
            world=self.subject.world
        )
        entity.is_ephemeral = True
        self.subject.world.children.append(entity)
        self.traces.append(entity)

    def _seconds_since(self, date: datetime) -> int:
        return abs(int((datetime.now() - date).total_seconds()))

    def trace_offset(self):
        return Point(0, 0)

    def remove_expired_traces(self):
        for trace in self.traces:
            if self._seconds_since(trace.creation_date) > self.trace_expiration_time:
                self.remove(trace)

    def remove(self, trace):
        trace.kill()
        self.traces.remove(trace)

    @staticmethod
    def next_trace_id():
        try:
            LeavesTracesWhileWalking.incremental_id += 1
        except AttributeError:
            LeavesTracesWhileWalking.incremental_id = 1
        return f'Trace-{LeavesTracesWhileWalking.incremental_id}'
